_posts/2024-05-31-analyzing-railway-traffic-in-the-netherlands.md

```diff
@@ -0,0 +1,417 @@
+GROUP BY month;
+```
+
+| month |      station       | num_services |
+|------:|--------------------|-------------:|
+| 1     | Utrecht Centraal   | 34760        |
+| 2     | Utrecht Centraal   | 32300        |
+| 3     | Utrecht Centraal   | 37386        |
+| 4     | Amsterdam Centraal | 33426        |
+| 5     | Utrecht Centraal   | 35383        |
+| 6     | Utrecht Centraal   | 35632        |
+
+Maybe surprisingly, in most months, the busiest railway station is not in Amsterdam but in the country's 4th largest city, [Utrecht](https://en.wikipedia.org/wiki/Utrecht), thanks to its central geographic location.
+
+## Finding the Top-3 Busiest Stations for Each Summer Month
+
+Let's change the question to: _Which are the top-3 busiest stations for each summer month?_
+The `arg_max()` function only helps us find the top-1 value but it is not sufficient for finding top-k results.
+Luckily, DuckDB has extensive support for SQL features, including [window functions](/docs/sql/window_functions) and we can use the [`rank()` function](/docs/sql/window_functions#rank) to find top-k values:
+
+```sql
+SELECT month, month_name, array_agg(station) AS top3_stations
+FROM (
+    SELECT
+        month,
+        strftime(make_date(2023, month, 1), '%B') AS month_name,
+        rank() OVER
+            (PARTITION BY month ORDER BY num_services DESC) AS rank,
+        station,
+        num_services
+    FROM services_per_month
+    WHERE month BETWEEN 6 AND 8
+)
+WHERE rank <= 3
+GROUP BY ALL
+ORDER BY month;
+```
+
+This gives the following result:
+
+| month | month_name |                        top3_stations                         |
+|------:|------------|--------------------------------------------------------------|
+| 6     | June       | [Utrecht Centraal, Amsterdam Centraal, Schiphol Airport]     |
+| 7     | July       | [Utrecht Centraal, Amsterdam Centraal, Schiphol Airport]     |
+| 8     | August     | [Utrecht Centraal, Amsterdam Centraal, Amsterdam Sloterdijk] |
+
+We can see that the top-3 spots are shared between four stations: Utrecht Centraal, Amsterdam Centraal, Schiphol Airport, and Amsterdam Sloterdijk.
+
+### Directly Querying Parquet Files through HTTPS or S3
+
+<!--
+    core features used: GROUP BY ALL
+    extensions used: httpfs, parquet
+-->
+
+DuckDB supports directly querying files, including CSV and Parquet, via [the HTTP(S) protocol and the S3 API](/docs/extensions/httpfs).
+For example, we can run the following query:
+
+```sql
+SELECT "Service:Date", "Stop:station name"
+FROM 'https://blobs.duckdb.org/data/services-2023.parquet'
+LIMIT 3;
+```
+
+It returns the following result:
+
+| Service:Date | Stop:Station name  |
+|--------------|--------------------|
+| 2023-01-01   | Rotterdam Centraal |
+| 2023-01-01   | Delft              |
+| 2023-01-01   | Den Haag HS        |
+
+> For querying private S3 buckets, you can use DuckDB's [Secrets Manager](/docs/extensions/httpfs/s3api#configuration-and-authentication) to authenticate.
```

Comment:
Do we want to include this?
We have another blog post covering the secret manager and it's not used here, I feel it's a little out of place here